File: frontend/server/tools/bash-tool.js
import { tool } from '@openai/agents';
import { z } from 'zod';
import { spawn } from 'child_process';
import path from 'path';

/**
 * Core bash execution function
 */
export const executeBashCommand = async ({ command, cwd = '/tmp' }) => {
  return new Promise((resolve) => {
    console.log(`[BASH] Executing: ${command}`);
    console.log(`[BASH] Working directory: ${cwd}`);
    
    // Safety checks
    const dangerousPatterns = [
      /rm\s+-rf\s+\//,  // rm -rf /
      /:\(\)\{.*\|:&\};:/,  // Fork bomb
      /mkfs/,  // Formatting commands
      /dd\s+if=.*of=\/dev/,  // Direct disk writes
    ];
    
    for (const pattern of dangerousPatterns) {
      if (pattern.test(command)) {
        resolve({
          success: false,
          error: 'Command blocked for safety reasons',
          stdout: '',
          stderr: 'This command pattern is not allowed'
        });
        return;
      }
    }
    
    // Execute command
    const proc = spawn('bash', ['-c', command], {
      cwd,
      env: {
        ...process.env,
        // Ensure Python tools have access to Shopify credentials
        SHOPIFY_SHOP_URL: process.env.SHOPIFY_SHOP_URL,
        SHOPIFY_ACCESS_TOKEN: process.env.SHOPIFY_ACCESS_TOKEN,
        OPENAI_API_KEY: process.env.OPENAI_API_KEY,
        PERPLEXITY_API_KEY: process.env.PERPLEXITY_API_KEY,
      },
      // Capture output
      shell: false
    });
    
    let stdout = '';
    let stderr = '';
    
    proc.stdout.on('data', (data) => {
      stdout += data.toString();
    });
    
    proc.stderr.on('data', (data) => {
      stderr += data.toString();
    });
    
    proc.on('close', (code) => {
      const result = {
        success: code === 0,
        exitCode: code,
        stdout: stdout.trim(),
        stderr: stderr.trim()
      };
      
      console.log(`[BASH] Exit code: ${code}`);
      if (stdout) console.log(`[BASH] stdout: ${stdout.substring(0, 200)}${stdout.length > 200 ? '...' : ''}`);
      if (stderr) console.log(`[BASH] stderr: ${stderr.substring(0, 200)}${stderr.length > 200 ? '...' : ''}`);
      
      // Format result for agent
      if (code === 0) {
        resolve(stdout || 'Command executed successfully (no output)');
      } else {
        resolve(`Command failed with exit code ${code}\n${stderr || stdout || 'No error output'}`);
      }
    });
    
    proc.on('error', (err) => {
      console.error(`[BASH] Process error:`, err);
      resolve(`Failed to execute command: ${err.message}`);
    });
    
    // Timeout after 30 seconds
    setTimeout(() => {
      proc.kill();
      resolve('Command timed out after 30 seconds');
    }, 30000);
  });
};

/**
 * Bash tool for agents - provides controlled command line access
 */
export const bashTool = tool({
  name: 'bash',
  description: `Execute bash commands. You have access to:
    - Python tools in /home/pranav/idc/tools/ and /home/pranav/espressobot/frontend/server/python-tools/
    - Standard Unix utilities (grep, awk, sed, jq, etc.)
    - Python 3 with all Shopify/e-commerce libraries installed
    - Can read/write temporary files in /tmp/
    
    Examples:
    - python3 /home/pranav/idc/tools/search_products.py "coffee" --status active
    - python3 /home/pranav/idc/tools/get_product.py SKU123 | jq '.price'
    - echo "SKU123,49.99" > /tmp/price_updates.csv && python3 /home/pranav/idc/tools/bulk_price_update.py /tmp/price_updates.csv
    
    Safety notes:
    - Always use absolute paths for tools
    - Be careful with quotes and escaping
    - Check command output for errors`,
  parameters: z.object({
    command: z.string().describe('The bash command to execute'),
    cwd: z.string().nullable().default('/tmp').describe('Working directory (defaults to /tmp)')
  }),
  execute: executeBashCommand
});

/**
 * Create a bash-enabled agent
 */
export function createBashAgent(name, task) {
  const { Agent } = require('@openai/agents');
  
  return new Agent({
    name,
    instructions: `You are a bash-enabled agent. Your task: ${task}
    
    You have full bash access with Python tools available at:
    - /home/pranav/idc/tools/ (main tools)
    - /home/pranav/espressobot/frontend/server/python-tools/ (additional tools)
    
    Best practices:
    1. Always check if a tool exists before using it: ls -la /path/to/tool.py
    2. Use --help to understand tool parameters: python3 tool.py --help
    3. Check exit codes and stderr for errors
    4. Use temporary files in /tmp/ when needed
    5. Chain commands with && for sequential operations
    6. Use jq for JSON parsing when needed
    
    Example workflow:
    - First explore available tools: ls /home/pranav/idc/tools/*.py
    - Check tool usage: python3 /home/pranav/idc/tools/search_products.py --help
    - Execute tool: python3 /home/pranav/idc/tools/search_products.py "query" --status active
    - Process results: ... | jq '.[] | select(.vendor == "Test")'`,
    tools: [bashTool],
    model: 'gpt-4.1-mini'
  });
}

File: langgraph-backend/app/tools/bash_tool.py
"""
Bash execution tool for the bash agent.
Executes bash commands in the project's root directory.
"""

import os
import asyncio
import subprocess
import tempfile
from pathlib import Path
from typing import Dict, Any, Optional
import logging

logger = logging.getLogger(__name__)

class BashTool:
    def __init__(self):
        # Set up the root directory
        self.root_dir = Path(__file__).parent.parent.parent 
        logger.info(f"Bash tool initialized with root directory: {self.root_dir}")
    
    async def execute_command(self, command: str, timeout: int = 30) -> Dict[str, Any]:
        """
        Execute a bash command in the project's root directory.

        Args:
            command: The bash command to execute
            timeout: Timeout in seconds (default: 30)

        Returns:
            Dict with output, error, and return_code
        """
        try:
            # Log the command being executed
            logger.info(f"Executing bash command: {command}")

            # Execute command in the root directory
            process = await asyncio.create_subprocess_shell(
                command,
                cwd=str(self.root_dir),
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
                env={**os.environ, 'PWD': str(self.root_dir)}
            )

            try:
                stdout, stderr = await asyncio.wait_for(
                    process.communicate(),
                    timeout=timeout
                )
                return_code = process.returncode
            except asyncio.TimeoutError:
                process.kill()
                await process.wait()
                return {
                    "success": False,
                    "error": f"Command timed out after {timeout} seconds",
                    "return_code": -1,
                    "stdout": "",
                    "stderr": "",
                }

            # Decode output
            stdout_str = stdout.decode('utf-8', errors='replace') if stdout else ""
            stderr_str = stderr.decode('utf-8', errors='replace') if stderr else ""

            success = return_code == 0

            result = {
                "success": success,
                "return_code": return_code,
                "stdout": stdout_str,
                "stderr": stderr_str,
                "working_directory": str(self.root_dir)
            }

            if not success:
                result["error"] = f"Command failed with return code {return_code}"
                if stderr_str:
                    result["error"] += f": {stderr_str}"

            logger.info(f"Command completed: success={success}, return_code={return_code}")
            return result

        except Exception as e:
            logger.error(f"Error executing bash command: {e}")
            return {
                "success": False,
                "error": f"Execution error: {str(e)}",
                "return_code": -1,
                "stdout": "",
                "stderr": "",
            }
    


# Global instance
bash_tool = BashTool()